src/units/abilities.cpp

```diff
@@ -469,32 +469,46 @@ bool unit::ability_active_impl(const std::string& ability,const config& cfg,cons
 
 	const unit_map& units = get_unit_map();
 
-	for (const config &i : cfg.child_range("filter_adjacent"))
-	{
+	for(const config &i : cfg.child_range("filter_adjacent")) {
+		std::size_t radius = i["radius"].to_int(1);
 		std::size_t count = 0;
 		unit_filter ufilt{ vconfig(i) };
 		ufilt.set_use_flat_tod(illuminates);
-		std::vector<map_location::direction> dirs = i["adjacent"].empty() ? map_location::all_directions() : map_location::parse_directions(i["adjacent"]);
-		for (const map_location::direction index : dirs)
-		{
-			unit_map::const_iterator unit = units.find(adjacent[static_cast<int>(index)]);
-			if (unit == units.end())
-				continue;
-			if (!ufilt(*unit, *this))
-				continue;
-			if((*this).id() == (*unit).id())
+		for(const unit& u : units) {
+			const map_location& from_loc = u.get_location();
+			std::size_t distance = distance_between(from_loc, loc);
+			if(distance > radius || !ufilt(u, *this)) {
 				continue;
-			if (i.has_attribute("is_enemy")) {
+			}
+			int dir = 0;
+			for(unsigned j = 0; j < adjacent.size(); ++j) {
+				bool adj_or_dist = distance != 1 ? distance_between(adjacent[j], from_loc) == (distance - 1) : adjacent[j] == from_loc;
+				if(adj_or_dist) {
+					dir = j;
+					break;
+				}
+			}
+			assert(dir >= 0 && dir <= 5);
+			map_location::direction direction{ dir };
+			if(i.has_attribute("adjacent")) { //key adjacent defined
+				std::vector<map_location::direction> dirs = map_location::parse_directions(i["adjacent"]);
+				if(std::find(dirs.begin(), dirs.end(), direction) == dirs.end()) {
```

Comment:
Small change, but these two lines can be simplified to:
```cpp
if(!utils::contains(map_location::parse_directions(i["adjacent"]), direction)) {
	continue;
}
```
No need for the `dirs` variable.